invisible_cities/cities/detsim_waveforms.py

```diff
@@ -0,0 +1,53 @@
+import numpy as np
+
+##################################
+######### WAVEFORMS ##############
+##################################
+def bincounter(xs : np.array,
+               dx : float =1.,
+               x0 : float =0.):
+    ixs = ((xs - x0) // dx).astype(int)
+    return np.unique(ixs, return_counts=True)
+
+
+def create_waveform(times : np.array,
+                    pes   : np.array,
+                    bins  : np.array,
+                    wf_bin_time : float,
+                    nsamples    : int) -> np.array:
+    wf = np.zeros(len(bins))
+
+    if np.sum(pes)==0:
+        return wf
+    if nsamples==0:
+        sel = pes>0
+        t = np.repeat(times[sel], pes[sel])
+        t = np.clip  (t, 0, bins[-1])
+        indexes, counts = bincounter(t, wf_bin_time)
+        wf[indexes] = counts
+        return wf
+
+    sel = pes>0
+    t = np.repeat(times[sel], pes[sel])
+    t = np.clip  (t, 0, bins[-nsamples])
+    indexes, counts = bincounter(t, wf_bin_time)
+
+    spread_pes = np.repeat(counts[:, np.newaxis]/nsamples, nsamples, axis=1)
+    for index, counts in zip(indexes, spread_pes):
+        wf[index:index+nsamples] = wf[index:index+nsamples] + counts        
```

Comment:
for consistency, rename `spread_pes` to `spread_counts`

Reply:
you can also use `+=` for incrementing the counts.

Reply:
What happens if `counts = 5` and `nsamples = 2`? You add 2.5 pes to each bin?
There is a risk of adding (or removing) counts unless the Poisson line in `create_sensor_waveforms` is managed properly, which might be the case, but maybe is good idea to clarify or leave some inline comment in the code if that is the case.

Reply:
If I understand your last comment, there should be no problem with the poisson since for X~Poiss(a) and Y~Poiss(b), then X+Y~Poiss(a+b). This is one of the reasons to do not apply poisson until the end in `create_sensor_waveforms`. The other reason is because the function could be accurately tested if no poisson is applied.

Reply:
No, my concern was mostly about the meaning of things and the connection with other parts of the code.

Since the Poisson randomization seems to be optional (it is an argument to `create_sensor_waveforms`), I would say that the output of `create_waveform` means "exact number of photons in these time bins". Essentially because in a simulation I don't think it makes sense to say you received 2.5 photons.

Now, assuming this is _not_ necessarily correct, then there must be a connection between `create_waveform` and applying the Poisson sampling in `create_sensor_waveforms`, otherwise, I don't understand it. In this case, I assume the output of `create_waveform` can mean "_average_ number of photons in these time bins" and then we can take that average and use it as the parameter of a Poisson distribution to get the actual number of photons detected. This makes sense to me, but it also introduces a subtle connection between one function and the other.

It can also be that you decided to use it both ways:
- If `nsamples = 1` (or 0 or whatever) then the output of `create_waveform` is directly the number of photons detected.
- Otherwise, if `nsamples > 1`, the output of `create_waveform` is not the number of photons detected, but the average, and to get the actual number of photons detected we need to sample a Poisson distribution.

If this is the case, that is OK, but it needs to be documented in detail.
Also, if this code is meant to work as I described it, then `poisson` is not an independent argument, as it is determined from the value of `nsamples`. But this is an extrapolation from what I think is going on, which may not be correct.

Reply:
Your point is correct, I have two proposals for `create_waveform`:

Note that `pes` values must be a vector of integers as it is the function now. In detsim, pes from an emission point will come from the product of `pes = #photons x LT`, where `#photons` is the number of photons emited in an emission point and `LT` the Light Table probability for that emission point. Since the `LT`probability is a float, `pes` could be a non integer.

1) We could convert `pes` values to integer inside the function and explicitly comment this as well as the fact that for `nsamples>1` the counts could be no integer. 

2) We could accept no integer values in `pes` and change the histogram procedure using `np.histogram` with `pes` as weights. 

I haven't probe it, but I think the 1) is better in performance since a `np.histagram` could be time consuming.

Reply:
If I understand correctly, 1 is not correct. If by convert you mean `pes = int(pes)`, then that is not correct, because the original `pes` means "_average_ number of photons" and therefore it needs to be "poissonized". 2. Is more correct and I seriously doubt that `np.histogramp` has performance issues (because essentially does the same as you, but in C).

Reply:
You are right, `pes` is "poissonized" in `pes_at_sensor` function in #697 . So I would keep the function as it is, but explicitly adding a comment for `nsamples>1`. Would this be enought?

Reply:
Mmm, I think poissonization should be the last step no? in the end we are producing waveforms with integer number of photoelectrons

Reply:
Maybe I am confusing something. As it is now, we poissonize in `pes_at_sensors` function first (in module detsim_simulate_photons in #697). So the argument variable `pes_at_sensors` in `create_sensor_waveforms` is an array of integers.
If we remove this poissonization we are in 1 above, where photoelectrons are  no necessarily integers.
Probably this poissonization is not necessary and we could just simply have `pes=int(#photons x LT)`.

Then, another poissonization is done at the end of `create_sensor_waveforms` if poisson argument is set to True. In this case the output waveforms have an integer number of photoelectrons as you say. I do not understand your last comment since this second poissonization is done at the end.

Reply:
> Probably this poissonization is not necessary and we could just simply have pes=int(#photons x LT).

No, this is physically incorrect, particularly when the average number of photons detected is very low. The poissonization **must** be done.

We have for sure to poissonize the number of photons arriving to a sensor, but after that I don't think it makes sense to distribute those photons in multiple bins and then poissonize again.

> Then, another poissonization is done at the end of create_sensor_waveforms

This is maybe wrong, according to what I said before.

>  if poisson argument is set to True

Why is this optional?

Reply:
I have not followed any details, but doing Poisson smearing twice on same simulated data sounds very wrong to me (If this is actually what the code does).

Reply:
This was in @jahernando original code, which comes from @gonzaponte too. Strictly speaking, the poissonization was included in `create_waveform`.

In my opinion, before thinking better, the last poisson makes no sense since when you redistribute in `nsamples` you are summing up poisson variables (if the first poissonization was already done) which is a poisson itself, so "poissonizing" is not correct.

This was set as optional argument for two reasons: one because you can make a test for the no-poissonized function; second because I wasn't totally sure about this was correct.

Reply:
> the last poisson makes no sense

I also think that it is not correct, but there are other considerations.

When you think about the physics, there are a number of processes that have fluctuations (considering S2 only):
- When you have an energy deposition in the detector, you compute the number of ionization electrons (ie), this number is poissonized, because what you have as an estimator is the average number of ie produced.
- For each ie, you have to estimate the number of photons emitted (nphot), which is gaussianly distributed according to the EL gain and the Fano Factor.
- Given a value for nphot, you can compute the number of pes in a given sensor multiplying nphot by the probability from a light table or the parametrization. This number is also an average: the _average_ number of photons detected by that sensor. Therefore, in order to get the actual number of photons detected in a given sensor (npes) you need to poissonize, which you do in `pes_at_sensors`.
- Now, given that quantity you may need to redistribute them over a number of time bins (nsample), which you do in `create_waveform`. That's a simulation trick that you may want to implement but, if you do, you must keep the same number of photons, i.e., you shouldn't say you have npes/nsample photoelectrons in a time bin and poissonize again, even if statistically it is equivalent.

> one because you can make a test for the no-poissonized function

I understand the logic, but I think if you fix the random seed for your test you should be able to test the function without tricks of this kind. We can discuss it if you find troubles.